### ISSUE TITLE
ResBlock's stride

### ISSUE THREAD

**Issue by JoyHuYY1412** [kazuto1011/deeplab-pytorch#39]
I wonder why you set stride=2 when implement the 'layer3' as:
`self.add_module("layer3", _ResBlock(n_blocks[1], 256, 128, 512, 2, 1))`
Is there any reason to do that?

**Comment by JoyHuYY1412** [comment]
What's more, do you think that ASPP module in deeplabV3 does not need bias?
since I saw that in V3 you all use `_ConvBatchNormReLU` which does not has bias.

**Comment by kazuto1011** [comment]
There is no personal idea in the parts you pointed out.
1. The ```stride=2``` setting is default in ResNet building, especially for the last three bottleneck groups. In DeepLab v2, the last two are switched to dilated convolution. I have verified that by comparing my pytorch model with decomposed caffemodel which is released by the authors (```convert.py```).
2. Convolution op does not need bias if it followed by batch normalization (redundancy). And this is the default behavior in [tensorflow API](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/conv2d). See [original DeepLab v3 ASPP implementation](https://github.com/tensorflow/models/blob/master/research/deeplab/model.py#L383-L478).

**Comment by JoyHuYY1412** [comment]
Thank you! 
So if I want to train deeplabV3 for VOC2012, and I know deeplabV3 has batch-norm in every ASPP layer, will I load `deeplabv2_resnet101_COCO_init.pth`  and freeze the bn module using `model.module.scale.freeze_bn()` to finetune this model?

**Comment by kazuto1011** [comment]
You shouldn't freeze the BNs in ASPP, which are not trained. The scope of ```freeze_bn()``` is every bn in the model.

**Comment by JoyHuYY1412** [comment]
Thank you. So is it still necessary to freeze the BNs in layer1 to layer5 (layers before ASPP)?

**Comment by kazuto1011** [comment]
I think the partial training is kind of reasonable If you would follow the DeepLab v2 training protocol that initializes with the ```deeplabv2_resnet101_COCO_init.pth``` that is pre-trained on MSCOCO and freezes the trained batch norm. However, that's not the official way, the v3 paper says:

> Our added modules on top of ResNet all include batch normalization parameters [38], which we found important to be trained as well. Since large batch size is required to train batch normalization parameters, we employ output stride = 16 and compute the batch normalization statistics with a batch size of 16. The batch normalization parameters are trained with decay = 0.9997. After training on the trainaug set with 30K iterations and initial learning rate = 0.007, we then freeze batch normalization parameters, employ output stride = 8, and train on the official PASCAL VOC 2012 trainval set for another 30K iterations and smaller base learning rate = 0.001.

According to the official tensorflow codes, a scope of ```fine_tune_batch_norm``` flag includes every batch norm both in ResNet and ASPP. You need to initialize the model with ImageNet or/and MS-COCO params, iterate 30k times with finetuning all the batch norm, and then iterate another 30k times with fixed batch norm. Note that the pytorch batch norm does not support multi-GPU training.

**Comment by kazuto1011** [comment]
FYI: https://github.com/vacancy/Synchronized-BatchNorm-PyTorch

**Comment by JoyHuYY1412** [comment]
Thank you! I will read them thoroughly~ :D

**Comment by JoyHuYY1412** [comment]
One more thing, I think since you only train these params `def get_params(model, key):
    # For Dilated FCN
    if key == "1x":
        for m in model.named_modules():
            if "layer" in m[0]:
                if isinstance(m[1], nn.Conv2d):
                    for p in m[1].parameters():
                        yield p
    # For conv weight in the ASPP module
    if key == "10x":
        for m in model.named_modules():
            if "aspp" in m[0]:
                if isinstance(m[1], nn.Conv2d):
                    yield m[1].weight
    # For conv bias in the ASPP module
    if key == "20x":
        for m in model.named_modules():
            if "aspp" in m[0]:
                if isinstance(m[1], nn.Conv2d):
                    yield m[1].bias`
maybe you don't need to `model.module.scale.freeze_bn()` ?
Because if you want to train the BN, you have to put their parmas in the optimizer's params?
I don't know if I'm right. ☘️

**Comment by kazuto1011** [comment]
Yes, you're right. My codebase is just for v2 training.

**Comment by JoyHuYY1412** [comment]
I am still confused by when we should train the BNs and when to freeze them. I hope you can give me some advice.
For example, when you load the init model trained on MSCoco and fine-tune it to work for Coco or PASCAL VOC, you should train all the BNs, right? Does that mean we can freeze all the other parameters and just train the parameters of BNs for some epochs?

So actually every time we change the dataset, we should train the BNs?

But if we don't change the dataset, just change the task, do we need to train BNs? Or just freeze them, but train other parameters? But training other layers' conv may influence the output's distribution, too. So I am confused.

**Comment by kazuto1011** [comment]
> Does that mean we can freeze all the other parameters and just train the parameters of BNs for some epochs?

No. I meant to finetune the BNs as well as other layers. Basically, we want to finetune all the parameters simultaneously, however, the BN training takes much cost because it needs a large batch size for estimating stats. One more costly thing is how much layers switch to dilated convolution, which makes the result coarser or denser. That's why the v2 protocol simply keeps the ImageNet/MS-COCO BN stats as-is and finetunes only the Conv layers on Pascal VOC, using 1/8 sized logits (output_stride=8). To be honest, I'm not sure about MS-COCO pre-training whose code has not released.

> Or just freeze them, but train other parameters? But training other layers' conv may influence the output's distribution

I think that's true. Therefore, the v3 protocol has two phases as I mentioned above. The first 30k iterations are to jointly train both BNs and Convs with 1/16 downsampling (output_stride=16, coarse logits). The second 30k iterations are to re-train only the Convs with 1/8 downsampling (output_stride=8, dense logits). The trade-off between accuracy and memory occupancy.

I hope the above is correct :)

**Comment by JoyHuYY1412** [comment]
Thank You~ It helps a lot. 🌷
